### PR TITLE
BUG: TimedeltaIndex raising ValueError when slice indexing (#16637)

### LINKED ISSUE #16637: BUG: TimedeltaIndex raising ValueError when slice indexing
This is tangentially related to #14946.

#### Code Sample, a copy-pastable example if possible

```python
import pandas as pd

df = pd.DataFrame({'x': range(10)})
df.index = pd.to_timedelta(df.index, unit='s')

# Each of these lines work as intended
df.loc[df.index[0], 'x'] = 10
df.loc[df['x'] > 1, 'x'] = 20
df.iloc[[0, 2], 0] = 30
df.iloc[2:4, 0] = 40

# These lines fail however
df.loc[df.index[7:9], 'x'] = 50
df.loc[df.index[3:5], 'x'] = 60
```

#### Problem description

This behavior is broken since 0.19.1 and is still occurring with the latest version of pandas. The problem is pretty self-explanatory.

**ValueError: The truth value of a Series is ambiguous. Use a.empty, a.bool(), a.item(), a.any() or a.all().**

#### Expected Output

The output is self-explanatory.

It can be easily fixed by adding the following to core/indexes/timedeltas.py TimedeltaIndex:

```python
def get_loc(self, key, method=None, tolerance=None):
    if is_bool_indexer(key) or is_timedelta64_dtype(key):
        raise TypeError
```

#### Output of ``pd.show_versions()``

<details>
INSTALLED VERSIONS
------------------
commit: None
python: 3.6.0.final.0
python-bits: 64
OS: Linux
OS-release: 4.4.0-53-generic
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: None
LANG: en_CA.UTF-8

pandas: 0.18.1
nose: None
pip: 9.0.1
setuptools: 27.2.0
Cython: 0.25.2
numpy: 1.11.2
scipy: 0.18.1
statsmodels: None
xarray: None
IPython: 5.1.0
sphinx: None
patsy: None
dateutil: 2.6.0
pytz: 2016.10
blosc: None
bottleneck: None
tables: 3.3.0
numexpr: 2.6.1
matplotlib: 1.5.3
openpyxl: None
xlrd: 1.0.0
xlwt: None
xlsxwriter: None
lxml: None
bs4: 4.5.3
html5lib: None
httplib2: None
apiclient: None
sqlalchemy: 1.1.4
pymysql: None
psycopg2: 2.6.2 (dt dec pq3 ext lo64)
jinja2: None
boto: None
pandas_datareader: None

</details>

### PR DESCRIPTION
**Pull request by jdeschenes**
Before accepting this Pull request is there any other kind of usage of loc that might fail using deltatimeindex that I am not aware of? I could take some time to fix it once and for all.

closes #16637

 - [ x ] closes #16637
 - [ x ] tests added / passed
 - [ x ] passes ``git diff upstream/master --name-only -- '*.py' | flake8 --diff``
 - [ x ] whatsnew entry

### CODE REVIEW

**Comment by jreback** [comment]
https://github.com/pandas-dev/pandas/issues?q=is%3Aopen+is%3Aissue+label%3ATimedelta+label%3AIndexing

are the only timedelta / indexing issues AFAICT (aside from new  one)

File: pandas/tests/indexing/test_timedelta.py [hunk 1]
```diff
@@ -18,3 +18,20 @@ def test_boolean_indexing(self):
                                     index=pd.to_timedelta(range(10), unit='s'),
                                     columns=['x'])
             tm.assert_frame_equal(expected, result)
+
```
Comment (jreback) [inline pandas/tests/indexing/test_timedelta.py]: just make sure test is near other slicing testsing (may not actually be many).

**Comment by jreback** [comment]
lgtm (just a comment to check location of test). ping on green.

File: pandas/tests/indexing/test_timedelta.py [hunk 2]
```diff
@@ -13,7 +13,24 @@ def test_boolean_indexing(self):
...
                          [10, 10, 10, 3, 4, 5, 6, 7, 8, 9]]
         for cond, data in zip(conditions, expected_data):
-            result = df.assign(x=df.mask(cond, 10).astype('int64'))
+            result = df.assign(x=df.mask(cond, 10).astype(df['x'].dtype))
+            expected = pd.DataFrame(data,
+                                    index=pd.to_timedelta(range(10), unit='s'),
+                                    columns=['x'])
+            tm.assert_frame_equal(expected, result)
+
+    def test_slice_indexing(self):
```
Comment (jorisvandenbossche) [inline pandas/tests/indexing/test_timedelta.py]: this is not slice indexing, but list-indexing with another index, if I am correct ? 
If so, I would update the name (and also whatsnew file)

File: pandas/tests/indexing/test_timedelta.py [hunk 3]
```diff
@@ -13,7 +13,24 @@ def test_boolean_indexing(self):
                          [0, 1, 2, 10, 4, 5, 6, 7, 8, 9],
                          [10, 10, 10, 3, 4, 5, 6, 7, 8, 9]]
         for cond, data in zip(conditions, expected_data):
-            result = df.assign(x=df.mask(cond, 10).astype('int64'))
+            result = df.assign(x=df.mask(cond, 10).astype(df['x'].dtype))
```
Comment (jorisvandenbossche) [inline pandas/tests/indexing/test_timedelta.py]: Is this change related to the PR?

Comment (jdeschenes) [inline pandas/tests/indexing/test_timedelta.py]: Not at all. The circle CI unit test fails on a 32 bit machine.

File: pandas/tests/indexing/test_timedelta.py (hunk 2 above)
Comment (jdeschenes) [inline pandas/tests/indexing/test_timedelta.py]: It was meant to convey that you take a slice off the index.

File: pandas/tests/indexing/test_timedelta.py (hunk 3 above)
Comment (jdeschenes) [inline pandas/tests/indexing/test_timedelta.py]: It still fails even with that attempted fix. Travis-CI builds is giving me error entirely unrelated to the change I have done.

File: pandas/tests/indexing/test_timedelta.py (hunk 2 above)
Comment (jreback) [inline pandas/tests/indexing/test_timedelta.py]: slice is actual slicing; here you are passing a boolean mask that is the same shape (and aligned) with the input); mark this as a separate case

**Comment by jreback** [comment]
do u have a repo for isnull or is_timeselta64_dtype raising?

**Comment by jdeschenes** [comment]
Yeah... I am not sure about this change for the is_timedelta64_dtype. On one hand the behavior to me is bogus(Either a value is of a type or it's not). On the other hand, other projects may rely on this behavior(Why? I have no idea). I ran the test against the pandas suite and nothing broke.

I can make the change locally in get_loc, but I would rather fix the issue directly in the function. Let me know what you think.

File: pandas/tests/indexing/test_timedelta.py (hunk 2 above)
Comment (jdeschenes) [inline pandas/tests/indexing/test_timedelta.py]: Hi, in the new case we are not passing a boolean mask. Would you agree if I renamed the test to `test_list_like_indexing`?

Comment (jorisvandenbossche) [inline pandas/tests/indexing/test_timedelta.py]: yep, that is better

File: pandas/core/dtypes/common.py [hunk 4]
```diff
@@ -396,7 +396,10 @@ def is_timedelta64_dtype(arr_or_dtype):
 
     if arr_or_dtype is None:
         return False
-    tipo = _get_dtype_type(arr_or_dtype)
+    try:
+        tipo = _get_dtype_type(arr_or_dtype)
+    except ValueError:
```
Comment (jreback) [inline pandas/core/dtypes/common.py]: this is fine
but need tests that hit this

**Comment by jdeschenes** [comment]
@jreback @jorisvandenbossche Both failures are unrelated to the code I changed. I think it's ready to be merged at this point.

File: pandas/tests/indexing/test_timedelta.py [hunk 5]
```diff
@@ -16,5 +16,24 @@ def test_boolean_indexing(self):
...
-                                    columns=['x'])
+                                    columns=['x'],
+                                    dtype='int64')
+            tm.assert_frame_equal(expected, result)
+
+    def test_list_like_indexing(self):
+        # GH 16637
+        df = pd.DataFrame({'x': range(10)}, dtype="int64")
+        df.index = pd.to_timedelta(range(10), unit='s')
+
```
Comment (jreback) [inline pandas/tests/indexing/test_timedelta.py]: might be slightly more clear as a parametrized test

e.g.

```
@pytest.mark.parametrize("indexer, exp", [(0, ....), (slice(4, 8), ...), ([3, 5], ...)]
def test_list_like(self, indexer, exp):
    ....
```

File: pandas/tests/dtypes/test_common.py [hunk 6]
```diff
@@ -204,6 +204,8 @@ def test_is_timedelta64_dtype():
     assert com.is_timedelta64_dtype(np.timedelta64)
     assert com.is_timedelta64_dtype(pd.Series([], dtype="timedelta64[ns]"))
 
+    assert not com.is_timedelta64_dtype("0 days 00:00:00")
```
Comment (jreback) [inline pandas/tests/dtypes/test_common.py]: didn't you have 2 cases that trigger the exception? if so pls add the other one.

File: doc/source/whatsnew/v0.20.3.txt [hunk 7]
```diff
@@ -37,7 +37,7 @@ Performance Improvements
 Bug Fixes
 ~~~~~~~~~
 
-
+- Bug in ``TimedeltaIndex`` raising a ``ValueError`` when slice indexing with ``loc`` (:issue:`16637`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.20.3.txt]: Bug in a DataFrame/Series with a TimedeltaIndex when slice indexing

**Comment by jreback** [comment]
lgtm. pls rebase and push again (had some ci failures unrelated).

File: pandas/tests/indexing/test_timedelta.py (hunk 5 above)
Comment (jdeschenes) [inline pandas/tests/indexing/test_timedelta.py]: Ok, I am not familiar with this... I will take a look.

**Comment by jreback** [comment]
http://pandas.pydata.org/pandas-docs/stable/contributing.html#using-pytest

File: pandas/tests/dtypes/test_common.py (hunk 6 above)
Comment (jdeschenes) [inline pandas/tests/dtypes/test_common.py]: No that I am aware of. There is a separate issue I think were isnull would trigger an exception in the get_loc function. I don't see how that condition could be hit in the future.

**Comment by jdeschenes** [comment]
Hi,

I have noticed another issue closely related to this.

```python
df.loc[[0, 3], 'x'] = 20
```

This will fail with the ValueError. Instead, it should do `KeyError: 'None of [[0, 3]] are in the [index]'`

I will take a second look a bit later and see if I can fix this.

**Comment by jreback** [comment]
you can address that other error in another PR (or here) up 2 u

**Comment by jdeschenes** [comment]
I Will do it in a separate PR

**Comment by jreback** [comment]
@jdeschenes ok great. if you can create an issue would be great.

FYI travis is fixed, working on circle now. pytest-xdist new version broke some things.

File: doc/source/whatsnew/v0.20.3.txt [hunk 8]
```diff
@@ -36,11 +36,11 @@ Performance Improvements
 
 Bug Fixes
 ~~~~~~~~~
+- Bug in a DataFrame/Series with a TimedeltaIndex when slice indexing (:issue:`16637`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.20.3.txt]: use double-backticks on TimedeltaIndex

**Review by jreback — CHANGES_REQUESTED** [review]
if you can add a few tests on the is_timedelta64_dtype (in tests/dtypes/test_common.py)

otherwise lgtm.

**Comment by TomAugspurger** [comment]
@jdeschenes can you update? (some linting failures too I believe)

**Comment by jdeschenes** [comment]
I have added the backticks to the what's new.

For the test, I have added an extra case. I can more if you want, if you have a better idea of what you want to see there, let me know.

**Comment by TomAugspurger** [comment]
@jdeschenes that's what failed this build: https://travis-ci.org/pandas-dev/pandas/jobs/249668871#L1779

**Comment by jreback** [comment]
can you rebase

**Comment by jdeschenes** [comment]
Let me know if you need anything else!

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks!

**Pull request merged**